feat: Add `for_pinning` support for `recipe.yaml`

Hofer-Julian opened the pull request:
- Replaces and quotes `pin_subpackage` and `pin_compatible` as done with `meta.yaml`
- Adds tests for newly introduced functions
- Includes missing pieces for `cbc` support from graph PR
- Ignore ruff lint for long lines, since formatting is enforced anyway

beckermr reviewed (changes requested):
Looking good! Two comments.

Hofer-Julian commented on conda_forge_tick/utils.py:
```diff
@@ -585,6 +615,17 @@ def _parse_validated_recipes(
             "summary": about.get("summary"),
         }
     )
+    if run_exports := requirements.get("run_exports"):
```
Good point, I extracted this block into a function with comments.
Hope that makes things clearer.

Hofer-Julian commented on conda_forge_tick/utils.py:
```diff
@@ -435,17 +436,18 @@ def parse_recipe_yaml(
         )
 
 
 def parse_recipe_yaml_containerized(
```
Nice find. Fixed!

beckermr commented on conda_forge_tick/utils.py:
```diff
@@ -667,6 +715,42 @@ def _parse_validated_recipes(
     return _remove_none_values(parsed_recipes)
 
 
+def parse_recipe_yaml_requirements(requirements):
```
I'd suggest using an `_` at the start of the function name here in order to not create more APIs we have ot think about. This one is clearly supposed to be private.

Hofer-Julian replied:
Good point!

Done in https://github.com/regro/cf-scripts/pull/2878/commits/2716239b004f3bb74b0b8aa4d72c658932283eba

beckermr commented on conda_forge_tick/utils.py:
```diff
@@ -667,6 +715,42 @@ def _parse_validated_recipes(
+    When rendering the recipe by rattler build,
+    `requirements["run_exports"]["weak"]` gives a list looking like:
+    [
+        {
+          "pin_subpackage": {
+            "name": "slepc",
+            "lower_bound": "x.x.x.x.x.x",
+            "upper_bound": "x.x"
+          }
+        }
```
This syntax appears to have an extra `{...}`?

Hofer-Julian replied:
Double-checked, and the output of rattler-build is indeed structured that way

beckermr commented on conda_forge_tick/utils.py:
```diff
@@ -667,6 +715,42 @@ def _parse_validated_recipes(
+    if run_exports := requirements.get("run_exports"):
+        weak = run_exports.get("weak")
+        if isinstance(weak, list) and isinstance(weak[0], dict):
+            run_exports["weak"] = [value for entry in weak for value in entry.values()]
+
+        strong = run_exports.get("strong")
+        if isinstance(strong, list) and isinstance(strong[0], dict):
+            run_exports["strong"] = [
+                value for entry in strong for value in entry.values()
```
I would have expected `.keys()` here (and above) given the doc string? I also think this might interact weirdly with the parsing for pinning since it comes after and is undoing dicts.

Let's add a unit test for this function to ensure it works properly and help document it's behavior.

Hofer-Julian replied:
I've added three

Hofer-Julian commented:
As usual, your reviews are much appreciated @beckermr!
I expect that I will be able to address them on Monday.

Hofer-Julian commented:
> The code needs more tests, both unit and integration.
> 
> I think it'd be worth translating more of the recipes I test with [here](https://github.com/regro/cf-scripts/blob/master/tests/test_parse_munged_run_export.py).

Most of your comments should be addressed now.
I've added more comments, more unit tests, and logic to handle more edge cases.

I've also ported another recipe as you suggested. The others were too complicated to be ported by `conda-recipe-manager`. If there's a specific case that you want to be handled, let me know, and I will try to make a recipe for that.

beckermr commented on tests/test_recipe_yaml/slepc_weak_strong.yaml:
```diff
@@ -0,0 +1,42 @@
+  run:
+    - petsc
+    - suitesparse
+  run_exports:
+    strong:
+      - ${{ pin_subpackage('slepc', upper_bound='x') }}
+    weak:
+      - ${{ pin_subpackage('slepc', upper_bound='x.x') }}
```
The build strings and comments are important for this test as they catch an important edge case.

Hofer-Julian replied:
I've re-added the comments. However, I cannot add the build strings.
This combination of `pin_subpackage` and build strings is not supported by rattler-build

wolfv replied:
@beckermr - the way `pin_subpackage` and `pin_compatible` work in rattler-build it doesn't really allow any extra characters on the `pin_subpackage` line. We basically lazy evaluate the function at a later time in order to not have to do the crazy recursive eval of conda-build (which makes it slow). 

It might actually be something to test in `rattler-build`. I see two options to cover this case:

- we could add a new keyword argument to the function (`build: "foo_*"`)
- we could extend the parsing in a way that we parse the rest of the string as the build argument

I think the first of the two options would be much easier to implement. 

Maybe @h-vetinari has opinions because he triggered some larger re-shuffeling of the pin-function arguments :)

wolfv replied:
For more context (@h-vetinari): the current requirement looks something like:

```yaml
- ${{ pin_subpackage(foo, max_pin="x.x") real_*  # some comment
```

beckermr replied:
Ahhhh I see. We'll need something in the v2 spec for this as it is a use case we care about. Agreed that a `build` keyword is probably best.

h-vetinari replied:
Yeah, I think a new kwarg sounds like a good idea! The previous reshuffling was about adding expressiveness (and reducing confusion because things didn't work as might be expected with the possibility to ask for contradictory things), so being able to to handle a case like [slepc](https://github.com/conda-forge/slepc-feedstock/blob/fdc04765da95151ad337532c24d72dc46932ff81/recipe/meta.yaml#L41-L44) (pin on additional constraints, but not as tight as `exact=True`) without workarounds sounds like a good addition to me.

How would we write this? 
```
  run_exports:
    - if: cuda_compiler_version == "None"
      then: ${{ pin_subpackage('slepc', upper_bound='x.x', build=scalar ~ '_*') }}
      else: ${{ pin_subpackage('slepc', upper_bound='x.x', build='cuda' ~ cuda_major ~ '_*') }}
```
Both `scalar` & `cuda_major` are variables; one from CBC, one derived from `cuda_compiler_version`. I'm assuming they are in context for when the function call happens? Or can we define a variable that corresponds to the build string (depending on the various combinations) and then just do:
```
  run_exports:
    - ${{ pin_subpackage('slepc', upper_bound='x.x', build=some_variable) }}
```

I think repeated application of `- if X: then: Y` makes the recipe pretty unreadable, so I'd definitely want to avoid a quadruple-if just for that run-export. Sidenote: are multi-line `then:` & `else:` clauses possible? I'm assuming they pytorch example here is a mechanical translation, because 
```
        - if: "cuda_compiler_version != \"None\""
          then: cudnn
        - if: "cuda_compiler_version != \"None\""
          then: nccl
        - if: "cuda_compiler_version != \"None\""
          then: magma
```
should really be
```
        - if: cuda_compiler_version != "None"
          then: cudnn
                nccl      # how do I write this?
                magma
```

wolfv replied:
Yes, multi-line if-else – as you've typed – is possible. `conda-recipe-manager` doesn't do it nicely (for humans) yet.

Hofer-Julian replied:
Personally, I don't mind the discussion here, but I wonder if we should open a new issue for this at https://github.com/prefix-dev/rattler-build so we can find it again later on

wolfv replied:
I just created an issue: https://github.com/prefix-dev/rattler-build/issues/1012

beckermr reviewed (changes requested):
A few more nits. We should add one test with recipe outputs.

Hofer-Julian commented on conda_forge_tick/utils.py:
```diff
@@ -493,6 +495,9 @@ def parse_recipe_yaml_containerized(
         args += ["--arch", arch]
 
+    if cbc_path is not None:
+        args += ["--cbc-path", cbc_path]
```
I've added a test `test_parse_cudnn_recipe_yaml`

Hofer-Julian commented on conda_forge_tick/utils.py:
```diff
@@ -680,6 +728,58 @@ def _parse_validated_recipes(
+            result = []
+            for entry in original:
+                if isinstance(entry, str):
+                    result.append(entry)
```
Yes, it was an oversight that this line wasn't run. Fixed!

Hofer-Julian commented:
@beckermr summary of the commits that I added since your last review:
- Added test `test_parse_cudnn_recipe_yaml` to verify `cbc_path` works as expected
- Changed logic, so a line that should have been run is actually run
- Remove unused flags `arch` (part of platform with `rattler-build`) and `log_debug`

beckermr commented on docker/run_bot_task.py:
```diff
@@ -503,30 +501,22 @@ def parse_meta_yaml(
     type=str,
     default=None,
-    help="The platform (e.g., 'linux', 'osx', 'win').",
+    help="The platform (e.g., 'linux-64', 'osx-arm64', 'win-64').",
```
Just an FYI in conda platform is linux, osx, win and arch is x86_64 (aka 64), arm64, ppc64le, etc. IDK what rattler-build is doing, but the doc string and usage here is confusing. I'd rename this option to `--platform-arch` if that is how it is supposed to be used.

beckermr replied:
We should also double check the code to ensure the correct value is being passed here.

beckermr commented on conda_forge_tick/utils.py:
```diff
@@ -524,9 +509,7 @@ def parse_recipe_yaml_local(
+        The platform (e.g., 'linux-64', 'osx-arm64', 'win-64').
     cbc_path : str, optional
         The path to global pinning file.
     log_debug : bool, optional
```
The doc strings need to be updated for removed options.

beckermr commented on conda_forge_tick/utils.py:
```diff
@@ -524,9 +509,7 @@ def parse_recipe_yaml_local(
-        The platform (e.g., 'linux', 'osx', 'win').
-    arch : str, optional
-        The CPU architecture (e.g., '64', 'aarch64').
+        The platform (e.g., 'linux-64', 'osx-arm64', 'win-64').
```
See my comment on platform vs arch below. We should call this one `platform_arch` if this is what we intend or similar.

beckermr reviewed (changes requested):
A few more items. Sorry for the slow drip of PR comments and review here. The bot code base may confuse platform vs arch in other spots. Let's fix this one and I will make an issue to check the rest of the repo.

Hofer-Julian commented:
@beckermr I've renamed everything to `platform_arch`.
I've also adapted one spot where `platform` instead of `platform_arch` was passed.
